File: sdk/typescript/src/lib/handlers/TagHandler.ts
import { MessageHandler } from './MessageHandler';
import { WebSocket } from 'ws';
import { WebSocketMessage } from '../WebSocketRouter';
import { LibraryServerDataSQLite } from '../LibraryServerDataSQLite';

export class TagHandler extends MessageHandler {
  constructor(
    dbService: LibraryServerDataSQLite,
    ws: WebSocket,
    message: WebSocketMessage
  ) {
    super(dbService, ws, message);
  }

  async handle(): Promise<void> {
    try {
      const { action, payload } = this.message;
      const { data } = payload;
      
      let result;
      switch(action) {
        case 'query':
          result = await this.dbService.queryTag(data.query);
          break;
        case 'create':
          result = await this.dbService.createTag(data);
          break;
        case 'update':
          result = await this.dbService.updateTag(data.id, data);
          break;
        case 'delete':
          result = await this.dbService.deleteTag(data.id);
          break;
        default:
          throw new Error(`Unsupported tag action: ${action}`);
      }

      this.sendResponse({data: result});
    } catch (err) {
      this.sendError(err instanceof Error ? err.message : 'Tag operation failed');
    }
  }
}

File: lib/plugins/libraries/services/websocket_handlers/tag_handler.dart
import 'package:mira/plugins/libraries/services/websocket_handlers/common_handler.dart';

class TagHandler extends BaseWebSocketHandler {
  final String action;
  final Map<String, dynamic> payload;

  TagHandler({
    required super.dbService,
    required super.channel,
    required super.data,
    required super.requestId,
    required super.libraryId,
    required this.action,
    required this.payload,
  });

  @override
  Future<void> handle() async {
    try {
      switch (action) {
        case 'all':
          return sendSuccess({'tags': await dbService.getAllTags()});
        case 'read':
          if (payload.containsKey('id')) {
            final record = await dbService.getTag(payload['id'] as int);
            sendSuccess({'record': record});
          } else {
            final records = await dbService.getTags(
              limit: payload['limit'] as int? ?? 100,
              offset: payload['offset'] as int? ?? 0,
            );
            sendSuccess({'records': records});
          }
          break;

        case 'create':
          final id = await dbService.createTag(data);
          final tags = await dbService.getAllTags();
          sendSuccess({'id': id, 'tags': tags});
          break;

        case 'update':
          final success = await dbService.updateTag(
            data['id'] as int,
            data['data'] as Map<String, dynamic>,
          );
          if (success) {
            final tags = await dbService.getAllTags();
            sendSuccess({'tags': tags});
          } else {
            sendError('Update failed');
          }
          break;

        case 'delete':
          final success = await dbService.deleteTag(data['id'] as int);
          if (success) {
            final tags = await dbService.getAllTags();
            sendSuccess({'tags': tags});
          } else {
            sendError('Delete failed');
          }
          break;

        case 'file_tag':
          if (action == 'update') {
            final success = await dbService.setFileTags(
              data['id'] as int,
              (data['data'] as List).cast<String>(),
            );
            sendSuccess({'success': success});
          } else if (action == 'read') {
            final tags = await dbService.getFileTags(data['id'] as int);
            sendSuccess({'tags': tags});
          }
          break;

        default:
          sendError('Unknown action for tags');
      }
    } catch (e) {
      sendError('Tag operation failed', e.toString());
    }
  }
}
